### DISCUSSION TITLE
Add environment objects and methods

### DISCUSSION THREAD

**Discussion by byeze** [suren-atoyan/monaco-react#242] [Q&A] [answered]
Hello!

Currently I'm doing a code based rule decision system, the thing is that I render a dashboard with a Monaco editor but I need to add the objects and methods passed to the function that you are coding.

For example, if I write: `user.`, I want the editor to show me suggestions like:
![image](https://user-images.githubusercontent.com/7424138/118062504-732bf580-b36d-11eb-948e-18fbdd0ee7ac.png)

How can I do it without automatically adding type definitions into the editor's value? 

Thank you for your help and time!

Edit: The editor allows you to write custom rules with javascript code that run on an isolated sandbox when the system needs to decide some action that the user requested, it's like when you write a custom rule on Auth0. The editor by default has this filled:
![image](https://user-images.githubusercontent.com/7424138/118067113-4fb97880-b376-11eb-9be3-d0db48581be2.png)
I need to suggest for example the user object when writing the rule.

**Comment by suren-atoyan** [comment]
Hello @byeze, sorry for the late answer.

You can use [completion-provider](https://microsoft.github.io/monaco-editor/playground.html#extending-language-services-completion-provider-example) to achieve that

**Comment by lucabiasotto** [comment, marked as answer]
Hi guys, have you found a solution to suggest properties based on the object?

I followed @suren-atoyan example but I don't understand how to make different suggestions based on the object.

For example, if I write `console` the monaco suggests `log`, `info` etc...i want that that if the user writes `mysql` monaco suggests my custom methods.

This below is my code, but it has the following 2 problems:

1. suggests `query` even though I haven't written `mysql`
![Schermata 2022-11-04 alle 14 33 54](https://user-images.githubusercontent.com/11436914/199985680-43dbc440-97e3-4af8-8dc0-7637ca7e3ab1.png)

2. when I select `mysql` from the suggestions then monaco doesn't suggest `query` (unless I press cmd + space)
 ![Schermata 2022-11-04 alle 14 38 20](https://user-images.githubusercontent.com/11436914/199986036-958df6ca-8e74-4738-a017-74c7675de012.png)


```javascript
monaco.languages.registerCompletionItemProvider("javascript", {
      provideCompletionItems: (model, position) => {
        var word = model.getWordUntilPosition(position);
        var range = {
          startLineNumber: position.lineNumber,
          endLineNumber: position.lineNumber,
          startColumn: word.startColumn,
          endColumn: word.endColumn
        };

        // Docs: https://microsoft.github.io/monaco-editor/api/interfaces/monaco.languages.CompletionItem.html
        return {
          suggestions: [
            {
              label: 'mysql',
              kind: monaco.languages.CompletionItemKind.Variable,// Module,
              documentation: "Anacleto library for working with MySQL.",
              commitCharacters: ["."],
              detail: "Anacleto MySQL library",
              insertText: 'mysql',
              range: range,
            },
            {
              label: 'query({})',
              filterText: 'mysql.query({})',
              kind: monaco.languages.CompletionItemKind.Function,
              insertTextRules: monaco.languages.CompletionItemInsertTextRule.InsertAsSnippet,
              documentation: "mysql.query: mysql.query <any>(query: string | timeout: int | values: array | nestTables: boolean)",
              detail: 'mysql.query({query..})',
              insertText: 'mysql.query({query:${1:string}, timeout:${2:millis}, values:${3:array}, nestTables:${4:boolean}})',
              range: range,
            }
          ]
        };
      }
    });
  }
```

**Reply by suren-atoyan** [reply]
please check [this](https://codesandbox.io/s/rough-haze-qcg3fz?file=/src/App.js). let me know if it helps

**Reply by lucabiasotto** [reply]
Thanks @suren-atoyan with this example it works!! Super :)
Bye
